### PR TITLE
Extract low-level store and pd crates.

### PR DESCRIPTION
**Pull request by ekexium**
Changes:

1. Extracting low-level kv-store and pd client crates could expose different levels of abstraction for to users for their interest. (#150 )
2. GetRequest now allows empty value.
3. Cargo.lock is removed from .gitignore because of weird build error for dependency codegen.
4. Batch test will skip batches with duplicated keys

### CODE REVIEW

File: src/mock.rs [hunk 1]
```diff
@@ -177,7 +177,7 @@ pub mod test {
     //     let addr2 = "bar";
     //
     //     let kv1 = client.kv_client(&addr1).unwrap();
-    //     let kv2 = client.kv_client(&addr2).unwrap();
+    //     let kv2 = client.kv_client(&addr2).unwrap();c
```
Comment (nrc) [inline src/mock.rs:180]: typo

File: Cargo.toml [hunk 2]
```diff
@@ -33,6 +33,10 @@ futures-timer = "3.0"
 async-trait = "0.1.27"
 tokio = { version = "0.2", features = ["sync"] }
 
+tikv-client-common = {path = "tikv-client-common"}
+tikv-client-pd = {path = "tikv-client-pd"}
+tikv-client-store = {path = "tikv-client-store"}
```
Comment (nrc) [inline Cargo.toml:38]: style nit: put a space before and after braces

File: src/mock.rs [hunk 3]
```diff
@@ -6,35 +6,35 @@
...
-    Config, Error, Key, Result,
+    Config, Error, Key, Result, Timestamp,
 };
 
+use tikv_client_pd::{PdClient, PdRpcClient, Region, RegionId, RetryClient, StoreBuilder};
+use tikv_client_store::{KvClient, KvConnect};
+
 use fail::fail_point;
+use futures::compat::Compat01As03;
 use futures::future::{ready, BoxFuture, FutureExt};
```
Comment (nrc) [inline src/mock.rs:18]: prefer to use the nested import style in the client

File: src/mock.rs [hunk 4]
```diff
@@ -6,35 +6,35 @@
...
-    pd::{PdClient, PdRpcClient, Region, RegionId, RetryClient},
     request::{DispatchHook, KvRequest},
-    transaction::Timestamp,
-    Config, Error, Key, Result,
+    Config, Error, Key, Result, Timestamp,
 };
 
+use tikv_client_pd::{PdClient, PdRpcClient, Region, RegionId, RetryClient, StoreBuilder};
+use tikv_client_store::{KvClient, KvConnect};
+
```
Comment (nrc) [inline src/mock.rs:15]: keep all imports together, not line-separated (sorry for the style nits)

File: src/mock.rs [hunk 5]
```diff
@@ -97,17 +108,13 @@ impl MockPdClient {
...
         region: Region,
-    ) -> BoxFuture<'static, Result<Store<Self::KvClient>>> {
-        Box::pin(ready(Ok(Store::new(
+    ) -> BoxFuture<'static, Result<StoreBuilder>> {
+        Box::pin(ready(Ok(StoreBuilder::new(
             region,
-            MockKvClient {
-                addr: String::new(),
-            },
+            String::new(),
```
Comment (nrc) [inline src/mock.rs:117]: Should this be a proper string?

File: src/raw/requests.rs [hunk 6]
```diff
@@ -81,19 +83,19 @@ impl KvRequest for kvrpcpb::RawBatchGetRequest {
...
-        let mut req = store.request::<Self>();
+        let mut req = self.request_from_store(store);
         req.set_keys(keys.into_iter().map(Into::into).collect());
         req.set_cf(self.cf.clone());
 
         req
     }
 
-    fn store_stream<PdC: PdClient>(
+    fn store_builder_stream<PdC: PdClient>(
```
Comment (nrc) [inline src/raw/requests.rs:93]: I think you can leave the name, "store_builder_stream" is a bit long and doesn't add much useful info

File: src/raw/requests.rs [hunk 7]
```diff
@@ -424,7 +426,8 @@ impl KvRequest for kvrpcpb::RawBatchScanRequest {
     const RPC_FN: RpcFnType<Self, Self::RpcResponse> = TikvClient::raw_batch_scan_async_opt;
 
     fn make_rpc_request<KvC: KvClient>(&self, ranges: Self::KeyData, store: &Store<KvC>) -> Self {
-        let mut req = store.request::<Self>();
+        // let mut req = store.request::<Self>();
```
Comment (nrc) [inline src/raw/requests.rs:429]: Can we remove the commented line?

File: src/transaction/lock.rs [hunk 8]
```diff
@@ -1,11 +1,13 @@
-use super::requests;
-use crate::pd::{PdClient, RegionVerId};
+// use crate::requests;
```
Comment (nrc) [inline src/transaction/lock.rs:1]: remove this comment too

File: tikv-client-pd/src/client.rs [hunk 9]
```diff
@@ -164,67 +175,61 @@ pub trait PdClient: Send + Sync + 'static {
 
 /// This client converts requests for the logical TiKV cluster into requests
 /// for a single TiKV store using PD and internal logic.
-pub struct PdRpcClient<KvC: KvConnect + Send + Sync + 'static = TikvConnect, Cl = Cluster> {
+pub struct PdRpcClient<Cl = Cluster> {
     pd: Arc<RetryClient<Cl>>,
-    kv_connect: KvC,
-    kv_client_cache: Arc<RwLock<HashMap<String, KvC::KvClient>>>,
+    // kv_connect: KvC,
+    // kv_client_cache: Arc<RwLock<HashMap<String, KvC::KvClient>>>,
```
Comment (nrc) [inline tikv-client-pd/src/client.rs:181]: remove the comments please

File: tikv-client-pd/src/client.rs [hunk 10]
```diff
@@ -164,67 +175,61 @@ pub trait PdClient: Send + Sync + 'static {
...
     }
 }
 
-impl<KvC: KvConnect + Send + Sync + 'static, Cl> PdRpcClient<KvC, Cl> {
-    pub async fn new<PdFut, MakeKvC, MakePd>(
+impl<Cl> PdRpcClient<Cl> {
+    pub async fn new<PdFut, MakePd>(
         config: &Config,
-        kv_connect: MakeKvC,
+        // kv_connect: MakeKvC,
```
Comment (nrc) [inline tikv-client-pd/src/client.rs:227]: and here

**Comment by nrc** [comment]
@sticnarf PTAL

File: src/proptests/mod.rs [hunk 11]
```diff
@@ -10,8 +10,8 @@ use std::env::var;
 mod raw;
 
 pub(crate) const ENV_PD_ADDRS: &str = "PD_ADDRS";
-pub(crate) const PROPTEST_KEY_MAX: usize = 1024 * 2; // 2 KB
-pub(crate) const PROPTEST_VALUE_MAX: usize = 1024 * 16; // 16 KB
+// pub(crate) const PROPTEST_KEY_MAX: usize = 1024 * 2; // 2 KB
+// pub(crate) const PROPTEST_VALUE_MAX: usize = 1024 * 16; // 16 KB
```
Comment (sticnarf) [inline src/proptests/mod.rs:14]: Can these unused consts just be removed?

**Comment by ekexium** [comment]
@sticnarf Comments resolved. PTAL

**Pull request merged**